File: WePeiYang/app/src/main/java/com/twt/service/settings/LibBindFragment.java
package com.twt.service.settings;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import com.twt.wepeiyang.commons.utils.CommonPrefUtil;
import com.twtstudio.retrox.auth.login.AuthSelfProvider;
import com.twtstudio.retrox.tjulibrary.provider.TjuLibProvider;
import com.twt.service.R;

import agency.tango.materialintroscreen.SlideFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * Created by retrox on 01/03/2017.
 */

public class LibBindFragment extends SlideFragment {

    private Unbinder unbinder;
    private boolean canMoveForward = false;
    private Context mContext;
    @BindView(R.id.lib_password)
    TextInputEditText libPasswordEdit;
    @BindView(R.id.btn_lib_bind)
    Button button;

    @OnClick(R.id.btn_lib_bind)
    public void bind_lib(View view) {

        String libPassWd = libPasswordEdit.getText().toString();
        if (TextUtils.isEmpty(libPassWd)) {
            libPassWd = "000000";
            Toast.makeText(mContext, "尝试默认密码ing...", Toast.LENGTH_SHORT).show();
        }
        new TjuLibProvider(this.getContext()).bindLibrary(integer -> {
            if (integer == -1) {
                canMoveForward = true;
                Toast.makeText(this.getContext(), "图书馆绑定完成,点击底部右侧对勾开始新旅程", Toast.LENGTH_SHORT).show();
            } else if (integer == 50003) {
                canMoveForward = true;
                Toast.makeText(this.getContext(), "图书馆已绑定,点击底部右侧对勾开始新旅程", Toast.LENGTH_SHORT).show();
            } else if (integer == 50002) {
                Toast.makeText(this.getContext(), "图书馆密码错误TAT...", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(this.getContext(), "未知错误TAT...", Toast.LENGTH_SHORT).show();
            }
            CommonPrefUtil.setIsBindLibrary(true);
            new AuthSelfProvider().getUserData();
        }, libPassWd);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_lib_bind_slide, container, false);
        canMoveForward = CommonPrefUtil.getIsBindLibrary();
        unbinder = ButterKnife.bind(this, view);
        mContext = this.getContext();
        return view;
    }

    @Override
    public int backgroundColor() {
        return R.color.white_color;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public int buttonsColor() {
        return R.color.intro_slide_buttons;
    }

    @Override
    public boolean canMoveFurther() {
        // TODO: 14/03/2017 强制绑定？？？
//        return this.canMoveForward;
        return true;
    }

    @Override
    public String cantMoveFurtherErrorMessage() {
        return "请绑定图书馆账号！";
    }
}

File: WePeiYang/app/src/main/java/com/twt/service/settings/LibBindFragment.kt
package com.twt.service.settings

import agency.tango.materialintroscreen.SlideFragment
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import butterknife.ButterKnife
import butterknife.Unbinder
import com.twt.service.R
import com.twt.wepeiyang.commons.experimental.cache.CacheIndicator.REMOTE
import com.twt.wepeiyang.commons.experimental.cache.RefreshState
import com.twt.wepeiyang.commons.experimental.preference.CommonPreferences
import com.twtstudio.retrox.auth.api.authSelfLiveData
import com.twtstudio.retrox.auth.api.login
import com.twtstudio.retrox.tjulibrary.provider.TjuLibProvider
import es.dmoral.toasty.Toasty

/**
 * Created by retrox on 01/03/2017.
 */

class LibBindFragment : SlideFragment() {

    private lateinit var libPasswordEdit: EditText
    private lateinit var button: Button
    private lateinit var unbinder: Unbinder

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_lib_bind_slide, container, false).also {
            unbinder = ButterKnife.bind(this, it)
        }
        libPasswordEdit = view.findViewById(R.id.lib_password)
        button = view.findViewById(R.id.btn_lib_bind)
        button.setOnClickListener {
            TjuLibProvider(context).bindLibrary({ integer ->
                when (integer) {
                    -1 -> {
                        /* 由于后台的接口问题， 每次绑定解绑操作都要重新拿token，干脆用login接口，增加用户体验*/
                        login (CommonPreferences.twtuname, CommonPreferences.password) {
                            when (it) {
                                is RefreshState.Success -> {
                                    authSelfLiveData.refresh(REMOTE)
                                }
                                is RefreshState.Failure -> {
                                    Toasty.error(context!!, "发生错误 ${it.throwable.message}！${it.javaClass.name}").show()
                                }
                            }
                        }
                        this.context?.let { it1 -> Toasty.success(it1, "图书馆绑定完成，点击底部右侧对勾开始新旅程", Toast.LENGTH_SHORT).show() }
                    }
                    50003 -> this.context?.let { it1 -> Toasty.success(it1, "图书馆已绑定，点击底部右侧对勾开始新旅程", Toast.LENGTH_SHORT).show() }
                    50002 -> this.context?.let { it1 -> Toasty.success(it1, "图书馆密码错误", Toast.LENGTH_SHORT).show() }
                    else -> this.context?.let { it1 -> Toasty.success(it1, "未知错误", Toast.LENGTH_SHORT).show() }
                }
            }, libPasswordEdit.text.toString().takeIf(String::isNotEmpty) ?: "000000")
        }
        return view
    }

    override fun backgroundColor(): Int = R.color.white_color

    override fun onDestroyView() {
        super.onDestroyView()
        unbinder.unbind()
    }

    override fun buttonsColor(): Int = R.color.intro_slide_buttons

    override fun canMoveFurther() = true

    override fun cantMoveFurtherErrorMessage() = "请绑定图书馆账号！"
}
